File: Java/04. Association Relationship/Practice14.java
/**
 * Practice 14: Smart Parking System
 * Task: Manage vehicle entry, slot assignment, payment, and exit.
 *       Calculate parking charge based on duration.
 *       Update slot allocation and availability in real time.
 *
 * How to compile and run:
 *   javac Practice14.java
 *   java Practice14
 *
 * Key Concepts:
 *   - Association between Vehicle, ParkingSlot, and ParkingLot
 *   - Duration-based charge calculation
 *   - Real-time availability tracking
 */

import java.time.LocalDateTime;
import java.time.Duration;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class Practice14 {

    /** Vehicle with license plate and type. */
    static class Vehicle {
        private String licensePlate;
        private String vehicleType;

        public Vehicle(String licensePlate, String vehicleType) {
            this.licensePlate = licensePlate;
            this.vehicleType = vehicleType;
        }

        public String getLicensePlate() { return licensePlate; }
        public String getVehicleType() { return vehicleType; }
    }

    /** ParkingSlot that can hold one vehicle at a time. */
    static class ParkingSlot {
        private String slotId;
        private String slotType;
        private double ratePerHour;
        private Vehicle vehicle;
        private LocalDateTime entryTime;

        public ParkingSlot(String slotId, String slotType, double ratePerHour) {
            this.slotId = slotId;
            this.slotType = slotType;
            this.ratePerHour = ratePerHour;
        }

        public boolean isAvailable() { return vehicle == null; }
        public String getSlotId() { return slotId; }
        public String getSlotType() { return slotType; }
        public double getRatePerHour() { return ratePerHour; }
        public Vehicle getVehicle() { return vehicle; }
        public LocalDateTime getEntryTime() { return entryTime; }

        public void assignVehicle(Vehicle vehicle, LocalDateTime entryTime) {
            this.vehicle = vehicle;
            this.entryTime = entryTime;
        }

        public void releaseVehicle() {
            this.vehicle = null;
            this.entryTime = null;
        }

        public void showInfo() {
            String status = isAvailable()
                    ? "Available"
                    : "Occupied by " + vehicle.getLicensePlate();
            System.out.printf("  %-8s | Type: %-10s | Rate: %.0f/hr | %s%n",
                    slotId, slotType, ratePerHour, status);
        }
    }

    /** ParkingLot manages a collection of ParkingSlots. */
    static class ParkingLot {
        private String name;
        private List<ParkingSlot> slots = new ArrayList<>();
        private static final DateTimeFormatter FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");

        public ParkingLot(String name) {
            this.name = name;
        }

        public void addSlot(ParkingSlot slot) {
            slots.add(slot);
        }

        public ParkingSlot findAvailableSlot(String slotType) {
            for (ParkingSlot slot : slots) {
                if (slot.getSlotType().equals(slotType) && slot.isAvailable()) {
                    return slot;
                }
            }
            return null;
        }

        public ParkingSlot parkVehicle(Vehicle vehicle, LocalDateTime entryTime) {
            ParkingSlot slot = findAvailableSlot(vehicle.getVehicleType());
            if (slot == null) {
                System.out.printf("  [Error] No available %s slot for %s.%n",
                        vehicle.getVehicleType(), vehicle.getLicensePlate());
                return null;
            }
            slot.assignVehicle(vehicle, entryTime);
            System.out.printf("  [OK] %s parked in slot %s at %s.%n",
                    vehicle.getLicensePlate(), slot.getSlotId(), entryTime.format(FMT));
            return slot;
        }

        public double exitVehicle(String licensePlate, LocalDateTime exitTime) {
            for (ParkingSlot slot : slots) {
                if (!slot.isAvailable() && slot.getVehicle().getLicensePlate().equals(licensePlate)) {
                    Duration duration = Duration.between(slot.getEntryTime(), exitTime);
                    double hours = duration.toMinutes() / 60.0;
                    if (hours < 1) hours = 1; // Minimum 1 hour
                    double charge = hours * slot.getRatePerHour();
                    System.out.printf("  [OK] %s exiting slot %s.%n", licensePlate, slot.getSlotId());
                    System.out.printf("       Duration : %.1f hours%n", hours);
                    System.out.printf("       Rate     : %.0f/hr%n", slot.getRatePerHour());
                    System.out.printf("       Charge   : %.2f%n", charge);
                    slot.releaseVehicle();
                    return charge;
                }
            }
            System.out.printf("  [Error] Vehicle %s not found in parking lot.%n", licensePlate);
            return 0;
        }

        public void showStatus() {
            long available = slots.stream().filter(ParkingSlot::isAvailable).count();
            long occupied = slots.size() - available;
            System.out.println("  Parking Lot: " + name);
            System.out.printf("  Total Slots: %d | Available: %d | Occupied: %d%n",
                    slots.size(), available, occupied);
            System.out.println();
            for (ParkingSlot slot : slots) {
                slot.showInfo();
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        ParkingLot lot = new ParkingLot("City Center Parking");
        lot.addSlot(new ParkingSlot("A-01", "Car", 50));
        lot.addSlot(new ParkingSlot("A-02", "Car", 50));
        lot.addSlot(new ParkingSlot("A-03", "Car", 50));
        lot.addSlot(new ParkingSlot("B-01", "Bike", 20));
        lot.addSlot(new ParkingSlot("B-02", "Bike", 20));
        lot.addSlot(new ParkingSlot("C-01", "Truck", 100));

        System.out.println("=== Initial Parking Status ===");
        lot.showStatus();

        LocalDateTime now = LocalDateTime.of(2026, 3, 7, 9, 0);
        System.out.println("=== Parking Vehicles ===");
        lot.parkVehicle(new Vehicle("DHK-1234", "Car"), now);
        lot.parkVehicle(new Vehicle("DHK-5678", "Car"), now.plusMinutes(15));
        lot.parkVehicle(new Vehicle("DHK-9012", "Bike"), now.plusMinutes(30));
        lot.parkVehicle(new Vehicle("CTG-3456", "Truck"), now.plusHours(1));
        System.out.println();

        System.out.println("=== After Parking ===");
        lot.showStatus();

        System.out.println("=== Vehicle Exits ===");
        lot.exitVehicle("DHK-1234", now.plusHours(3));
        System.out.println();
        lot.exitVehicle("DHK-9012", now.plusHours(1).plusMinutes(30));
        System.out.println();
        lot.exitVehicle("CTG-3456", now.plusHours(5));
        System.out.println();

        System.out.println("=== Final Parking Status ===");
        lot.showStatus();

        System.out.println("=== Parking When Slots Are Limited ===");
        lot.parkVehicle(new Vehicle("SYL-7777", "Car"), now.plusHours(4));
        lot.parkVehicle(new Vehicle("SYL-8888", "Car"), now.plusHours(4));
        lot.parkVehicle(new Vehicle("SYL-9999", "Car"), now.plusHours(4));
    }
}

File: Golang/04. Association Relationship/practice-14.go
/**
 * Practice 14: Smart Parking System
 * Task: Manage vehicle entry, slot assignment, payment, and exit.
 *       Calculate parking charge based on duration.
 *       Update slot allocation and availability in real time.
 *
 * How to run:
 *   go run practice-14.go
 *
 * Key Concepts:
 *   - Association between Vehicle, ParkingSlot, and ParkingLot
 *   - Duration-based charge calculation
 *   - Real-time availability tracking
 */

package main

import (
	"fmt"
	"time"
)

/** Vehicle with license plate and type. */
type Vehicle14 struct {
	LicensePlate string
	VehicleType  string
}

/** ParkingSlot that can hold one vehicle at a time. */
type ParkingSlot14 struct {
	SlotId      string
	SlotType    string
	RatePerHour float64
	Vehicle     *Vehicle14
	EntryTime   time.Time
}

func (s *ParkingSlot14) IsAvailable() bool {
	return s.Vehicle == nil
}

func (s *ParkingSlot14) ShowInfo() {
	status := "Available"
	if !s.IsAvailable() {
		status = "Occupied by " + s.Vehicle.LicensePlate
	}
	fmt.Printf("  %-8s | Type: %-10s | Rate: %.0f/hr | %s\n",
		s.SlotId, s.SlotType, s.RatePerHour, status)
}

/** ParkingLot manages a collection of ParkingSlots. */
type ParkingLot14 struct {
	Name  string
	Slots []*ParkingSlot14
}

func NewParkingLot14(name string) *ParkingLot14 {
	return &ParkingLot14{Name: name}
}

func (lot *ParkingLot14) AddSlot(slot *ParkingSlot14) {
	lot.Slots = append(lot.Slots, slot)
}

func (lot *ParkingLot14) FindAvailableSlot(slotType string) *ParkingSlot14 {
	for _, slot := range lot.Slots {
		if slot.SlotType == slotType && slot.IsAvailable() {
			return slot
		}
	}
	return nil
}

func (lot *ParkingLot14) ParkVehicle(vehicle *Vehicle14, entryTime time.Time) *ParkingSlot14 {
	slot := lot.FindAvailableSlot(vehicle.VehicleType)
	if slot == nil {
		fmt.Printf("  [Error] No available %s slot for %s.\n", vehicle.VehicleType, vehicle.LicensePlate)
		return nil
	}
	slot.Vehicle = vehicle
	slot.EntryTime = entryTime
	fmt.Printf("  [OK] %s parked in slot %s at %s.\n",
		vehicle.LicensePlate, slot.SlotId, entryTime.Format("2006-01-02 15:04"))
	return slot
}

func (lot *ParkingLot14) ExitVehicle(licensePlate string, exitTime time.Time) float64 {
	for _, slot := range lot.Slots {
		if !slot.IsAvailable() && slot.Vehicle.LicensePlate == licensePlate {
			duration := exitTime.Sub(slot.EntryTime)
			hours := duration.Hours()
			if hours < 1 {
				hours = 1
			}
			charge := hours * slot.RatePerHour
			fmt.Printf("  [OK] %s exiting slot %s.\n", licensePlate, slot.SlotId)
			fmt.Printf("       Duration : %.1f hours\n", hours)
			fmt.Printf("       Rate     : %.0f/hr\n", slot.RatePerHour)
			fmt.Printf("       Charge   : %.2f\n", charge)
			slot.Vehicle = nil
			slot.EntryTime = time.Time{}
			return charge
		}
	}
	fmt.Printf("  [Error] Vehicle %s not found in parking lot.\n", licensePlate)
	return 0
}

func (lot *ParkingLot14) ShowStatus() {
	available := 0
	for _, slot := range lot.Slots {
		if slot.IsAvailable() {
			available++
		}
	}
	occupied := len(lot.Slots) - available
	fmt.Println("  Parking Lot:", lot.Name)
	fmt.Printf("  Total Slots: %d | Available: %d | Occupied: %d\n", len(lot.Slots), available, occupied)
	fmt.Println()
	for _, slot := range lot.Slots {
		slot.ShowInfo()
	}
	fmt.Println()
}

func main() {
	lot := NewParkingLot14("City Center Parking")
	lot.AddSlot(&ParkingSlot14{SlotId: "A-01", SlotType: "Car", RatePerHour: 50})
	lot.AddSlot(&ParkingSlot14{SlotId: "A-02", SlotType: "Car", RatePerHour: 50})
	lot.AddSlot(&ParkingSlot14{SlotId: "A-03", SlotType: "Car", RatePerHour: 50})
	lot.AddSlot(&ParkingSlot14{SlotId: "B-01", SlotType: "Bike", RatePerHour: 20})
	lot.AddSlot(&ParkingSlot14{SlotId: "B-02", SlotType: "Bike", RatePerHour: 20})
	lot.AddSlot(&ParkingSlot14{SlotId: "C-01", SlotType: "Truck", RatePerHour: 100})

	fmt.Println("=== Initial Parking Status ===")
	lot.ShowStatus()

	now := time.Date(2026, 3, 7, 9, 0, 0, 0, time.Local)
	fmt.Println("=== Parking Vehicles ===")
	lot.ParkVehicle(&Vehicle14{"DHK-1234", "Car"}, now)
	lot.ParkVehicle(&Vehicle14{"DHK-5678", "Car"}, now.Add(15*time.Minute))
	lot.ParkVehicle(&Vehicle14{"DHK-9012", "Bike"}, now.Add(30*time.Minute))
	lot.ParkVehicle(&Vehicle14{"CTG-3456", "Truck"}, now.Add(1*time.Hour))
	fmt.Println()

	fmt.Println("=== After Parking ===")
	lot.ShowStatus()

	fmt.Println("=== Vehicle Exits ===")
	lot.ExitVehicle("DHK-1234", now.Add(3*time.Hour))
	fmt.Println()
	lot.ExitVehicle("DHK-9012", now.Add(90*time.Minute))
	fmt.Println()
	lot.ExitVehicle("CTG-3456", now.Add(5*time.Hour))
	fmt.Println()

	fmt.Println("=== Final Parking Status ===")
	lot.ShowStatus()

	fmt.Println("=== Parking When Slots Are Limited ===")
	lot.ParkVehicle(&Vehicle14{"SYL-7777", "Car"}, now.Add(4*time.Hour))
	lot.ParkVehicle(&Vehicle14{"SYL-8888", "Car"}, now.Add(4*time.Hour))
	lot.ParkVehicle(&Vehicle14{"SYL-9999", "Car"}, now.Add(4*time.Hour))
}
